### ISSUE TITLE
Using MQTT in DOODS

### ISSUE THREAD

**Issue by Idefix0496** [snowzach/doods2#42]
Hi,
I would like to know, how to use the MQTT Implementation in Doods. There is a **mqtt** section in the config.yaml file but no matter what I do I can't get Doods to connect to my broker.
Also I haven't found any documentation on that Subject. So do you have any suggestions what I'm doing wrong ?

**Comment by snowzach** [comment]
Did you figure it out? Where is your broker running? Same machine? Also in docker? They will need to be on the same network to communicate which might be your issue.

**Comment by dbeckers99** [comment]
I'm facing the same problem. My broker ist running on another machine in the same network. Can you please add more documentation or information about mqtt connection of Doods2? Thanks!

**Comment by dbeckers99** [comment]
Traceback (most recent call last):
  File "main.py", line 75, in <module>
    main()
  File "main.py", line 70, in main
    mqtt.run()
  File "/opt/doods/mqtt.py", line 104, in run
    self.mqtt_client.connect(self.config.broker.host, self.config.broker.port, 60)
  File "/usr/local/lib/python3.8/dist-packages/paho/mqtt/client.py", line 914, in connect
    return self.reconnect()
  File "/usr/local/lib/python3.8/dist-packages/paho/mqtt/client.py", line 1044, in reconnect
    sock = self._create_socket_connection()
  File "/usr/local/lib/python3.8/dist-packages/paho/mqtt/client.py", line 3685, in _create_socket_connection
    return socket.create_connection(addr, timeout=self._connect_timeout, source_address=source)
  File "/usr/lib/python3.8/socket.py", line 787, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/lib/python3.8/socket.py", line 918, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

**Comment by snowzach** [comment]
This sounds like a docker/networking config issue? What's your config look like?

**Comment by dbeckers99** [comment]
Thanks for reply!

I fixed the problem (correctly putting my config.yaml to docker container in run command), was my fault, sorry. mqtt works now for me.

A question: is it possible to publish the detection result picture to mqtt topic if requesting detection from rtsp stream? As far as I understand the code the detection result picture ist only published to images topic if the request was based on a single image (but not on a stream)

**Comment by bennyb0i** [comment]
Hey there,

I'm also curious about getting DOODS2 to work with MQTT. I've configured my mqtt broker in the config.yaml and copied it to the DOODS2 docker container, confirming that DOODS is loading the updated config (at least the non-MQTT portions) upon `docker start`.

Though, I don't see any indication in the logs that it's communicating with MQTT nor do I see any new topics published via MQTT Explorer. I can also confirm that the DOODS docker container can communicate with my MQTT broker via ping and that DOODS2 is working properly when sending a sample image to process using both the web test UI and POST via the RestAPI.

Any advice?

**Comment by dbeckers99** [comment]
It works with MQTT, but only with streaming data input, not with single images. Did you start the docker container with keyword mqtt?

docker run -v [local_path_to_yaml]/config.yaml:[docker_path_to_yaml]/config.yaml -e CONFIG_FILE=config.yaml -p 8080:8080 snowzach/doods2:latest **mqtt**

**Comment by bennyb0i** [comment]
Thanks, I started a new docker container using docker run and the mqtt flag you mentioned, but now I'm getting an error in the log that says:

`[tcp @ ox7f6e540039c0] Failed to resolve hostname cameraip: Name or service not known`

And when I post a request to DOODS, I also now get the error response:

`{"detail":"Not Found"}`

When I stop that container and run the container that doesn't have the mqtt keyword, I can POST fine without errors.

**Comment by bennyb0i** [comment]
Nevermind, I figured it out. I was getting the errors because I left

```
mqtt:
  data: rtsp://user:pass@cameraip:554/cam/realmonitor?channel=1&subtype=1
```

instead of changing it to my own rtsp stream.

On that note, @snowzach, any plans to consider mqtt support for single images as well?

**Comment by dbeckers99** [comment]
Good! But I figured out the stream processing with mqtt isn't stable. After some hours of running it produces many picture deciding errors and then stops running.

So I programmed a wrapper program with JS, fetching a single picture from my 3 cameras every few seconds, fetching the detection result (I only use person) and sending the result to my local mqtt broker, if it's a positive result.

Until now this works stable for some days.
